### PR TITLE
[fastlane_core] Convert value type of ConfigItems earlier

### PR DESCRIPTION
**Pull request by ohayon**
So that command line values (specifically from OneOff) are converted to their proper type correctly.

Example:

``` ruby
fastlane run changelog_from_git_commits commits_count:10
```

Now works correctly and the Integer 10 is provided to the action

This will fix all instances where actions are being called from the command line. It was prompted by this issue: https://github.com/fastlane/fastlane/issues/4862

### CODE REVIEW

**Comment by KrauseFx** [comment]
Some tests are failing. Also, could you provide a description on why this change is necessary?

File: fastlane_core/lib/fastlane_core/configuration/config_item.rb [hunk 1]
```diff
@@ -87,14 +87,12 @@ def valid?(value)
 
     # Returns an updated value type (if necessary)
     def auto_convert_value(value)
-      # Weird because of https://stackoverflow.com/questions/9537895/using-a-class-object-in-case-statement
-      case
-      when data_type == Array
+      if data_type == Integer
+        return value.to_i unless value.nil? || value.to_i.to_s != value.to_s
```
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/config_item.rb]: Uh, that's confusing and tricky regarding the priority of those things, can we refactor the `unless` and `||` to make it easier to read?

Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/config_item.rb]: Could we just use
```ruby
return value.to_i if value.to_i > 0
```
I think it will have the same results

File: fastlane_core/spec/configuration_spec.rb [hunk 2]
```diff
@@ -234,6 +235,18 @@
...
+        it "auto converts strings to integers" do
+          c = [
+            FastlaneCore::ConfigItem.new(key: :int_value,
+                                         type: Integer)
+          ]
+          config = FastlaneCore::Configuration.create(c, {
+            int_value: "10"
+          })
+
+          expect(config[:int_value]).to eq(10)
```
Comment (KrauseFx) [inline fastlane_core/spec/configuration_spec.rb]: Perfect ❤️

File: fastlane_core/lib/fastlane_core/configuration/config_item.rb (hunk 1 above)
Comment (ohwutup) [inline fastlane_core/lib/fastlane_core/configuration/config_item.rb]: We don't want to check for the value being greater than 0 because 0 could be a valid input depending on the action. We could refactor to break it into another line or two, but I believe this is the logic we need to use here.

File: fastlane_core/lib/fastlane_core/configuration/configuration.rb [hunk 3]
```diff
@@ -54,11 +54,11 @@ def verify_value_exists
       # Make sure the given value keys exist
       @values.each do |key, value|
         next if key == :trace # special treatment
-
         option = option_for_key(key)
         if option
+          @values[key] = option.auto_convert_value(value)
```
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/configuration.rb]: We're already calling `auto_convert_value` in `fetch`, right? But we overwrite the `@values[key]` value. Does that mean we do it twice? Can this lead to a problem?

**Comment by ohwutup** [comment]
Just added a brief description about why this is necessary. Tests are passing in CI.

File: fastlane_core/lib/fastlane_core/configuration/config_item.rb (hunk 1 above)
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/config_item.rb]: Good point, let's stick with the existing solution then, but please refactor to make the order more clear 👍

File: fastlane_core/lib/fastlane_core/configuration/configuration.rb (hunk 3 above)
Comment (asfalcone) [inline fastlane_core/lib/fastlane_core/configuration/configuration.rb]: We do end up doing this twice in some cases because of the different entry points to this code and the fact that values can be loaded into a `config` after it is initialized. This code we added will only be run on values that are present at initialize and later fetch will run on all values,[ but only if they don't already have a value](https://github.com/fastlane/fastlane/blob/master/fastlane_core/lib/fastlane_core/configuration/configuration.rb#L232) so I think this is OK. What do you think @KrauseFx ?

File: fastlane_core/lib/fastlane_core/configuration/config_item.rb [hunk 4]
```diff
@@ -87,14 +87,14 @@ def valid?(value)
 
     # Returns an updated value type (if necessary)
     def auto_convert_value(value)
-      # Weird because of https://stackoverflow.com/questions/9537895/using-a-class-object-in-case-statement
-      case
-      when data_type == Array
+      return nil if value.nil?
```
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/config_item.rb]: Nice 👍

File: fastlane_core/lib/fastlane_core/configuration/config_item.rb [hunk 5]
```diff
@@ -87,14 +87,14 @@ def valid?(value)
 
     # Returns an updated value type (if necessary)
     def auto_convert_value(value)
-      # Weird because of https://stackoverflow.com/questions/9537895/using-a-class-object-in-case-statement
```
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/config_item.rb]: Nice 👍

File: fastlane_core/lib/fastlane_core/configuration/configuration.rb (hunk 3 above)
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/configuration.rb]: Yeah, that should be fine 👍

File: fastlane_core/lib/fastlane_core/configuration/config_item.rb [hunk 6]
```diff
@@ -87,14 +87,14 @@ def valid?(value)
...
-      case
-      when data_type == Array
+      return nil if value.nil?
+
+      if data_type == Array
         return value.split(',') if value.kind_of?(String)
-      when data_type == Integer
-        return value.to_i unless value.nil?
```
Comment (KrauseFx) [inline fastlane_core/lib/fastlane_core/configuration/config_item.rb]: Nice!

File: fastlane_core/spec/configuration_spec.rb [hunk 7]
```diff
@@ -234,6 +235,18 @@
           expect(config[:false_value]).to eq(false)
           expect(config[:false_value2]).to eq(false)
         end
+
+        it "auto converts strings to integers" do
```
Comment (KrauseFx) [inline fastlane_core/spec/configuration_spec.rb]: Let's add another test for the value `"0"`, just because we mentioned it as an edge case before 👍

**Comment by KrauseFx** [comment]
I'd just like one more test, then it's ready from my side 👍

**Comment by KrauseFx** [comment]
Great improvement over the first iteration!

**Comment by asfalcone** [comment]
This is aweeeeesome 👍 🌸
